### PR TITLE
fix: Input batch to ShuffleRepartitioner.insert_batch should not be larger than configured batch size

### LINKED ISSUE #498: bug: org.apache.comet.CometNativeException: range end index 8221 out of range for slice of length 8192
### Describe the bug

It looks like we may have a hard-coded assumption in shuffle where we cannot process an incoming batch that is larger than the default batch size.

I have seen this error more than once but currently have a repro so will document this more as I investigate.

```
org.apache.comet.CometNativeException: range end index 8221 out of range for slice of length 8192
        ...
        at org.apache.comet.Native.executePlan(Native Method)
        at org.apache.comet.CometExecIterator.executeNative(CometExecIterator.scala:72)
        at org.apache.comet.CometExecIterator.getNextBatch(CometExecIterator.scala:129)
        at org.apache.comet.CometExecIterator.hasNext(CometExecIterator.scala:144)
        at org.apache.spark.sql.comet.execution.shuffle.CometShuffleWriteProcessor.write(CometShuffleExchangeExec.scala:492)
        at org.apache.spark.sql.comet.shims.ShimCometShuffleWriteProcessor.write(ShimCometShuffleWriteProcessor.scala:35)
        at org.apache.spark.sql.comet.shims.ShimCometShuffleWriteProcessor.write$(ShimCometShuffleWriteProcessor.scala:28)
        at org.apache.spark.sql.comet.execution.shuffle.CometShuffleWriteProcessor.write(CometShuffleExchangeExec.scala:448)
```

### Steps to reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #498.

### CODE REVIEW

**Comment by viirya** [comment]
cc @andygrove @sunchao

File: core/src/execution/datafusion/shuffle_writer.rs [hunk 1]
```diff
@@ -947,12 +948,18 @@ async fn external_shuffle(
...
     while let Some(batch) = input.next().await {
         let batch = batch?;
-        repartitioner.insert_batch(batch).await?;
+        let mut start = 0;
+        while start < batch.num_rows() {
+            let end = (start + batch_size).min(batch.num_rows());
+            let batch = batch.slice(start, end - start);
+            repartitioner.insert_batch(batch).await?;
+            start = end;
+        }
```
[the remark this answers is not in the document]
Comment (viirya) [reply core/src/execution/datafusion/shuffle_writer.rs]: Okay

**Comment by viirya** [comment]
cc @huaxingao @kazuyukitanimura

**Review by sunchao — APPROVED** [review]
LGTM

**Comment by viirya** [comment]
Thank you @sunchao

**Pull request merged**